### PR TITLE
Import pre implementation

### PR DESCRIPTION
**Pull request by k3rnel**

### CODE REVIEW

**Review by tiko-star — CHANGES_REQUESTED** [review]

File: app.js [hunk 1]
```diff
@@ -32,7 +33,10 @@ const emitter = new EventEmitter();
 const connectionInstance = DatabaseManager.getInstance("commerce");
 
 // After consuming start export of the document.
-emitter.on("consumed:export", onConsumed(emitter, connectionInstance));
+emitter.on("consumed:export", onExportConsumed(emitter, connectionInstance));
+
+// After consuming start export of the document.
```
[the remark this answers is not in the document]
Comment (tiko-star) [reply app.js:39]: Likel es u toxel nuyn dzev

File: config.js [hunk 2]
```diff
@@ -19,6 +19,7 @@ const {
     DB_DATABASE,
     DB_USERNAME,
     DB_PASSWORD,
+    ROOT_PATH
```
Comment (tiko-star) [inline config.js:22]: storaket dir misht verjum

File: handlers/onExportConsumed.js [hunk 3]
```diff
@@ -4,7 +4,7 @@ const config = require("../config.js");
 const DocumentCreator = require("../service/DocumentCreator");
 
 /**
- * Return event handler for emitted "consumed" events.
+ * Return event handler for emitted "exportConsumed" events.
```
Comment (tiko-star) [inline handlers/onExportConsumed.js:7]: eventi anuny consumed:export a

File: handlers/onExported.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ const config = require("../config.js");
  *
  * @returns {function} Event handler.
  */
-module.exports = function (compressionType) {
+module.exports = (compressionType) => {
```
Comment (tiko-star) [inline handlers/onExported.js:14]: es mi jnji

File: handlers/onImportConsumed.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * Return event handler for emitted "importConsumed" events.
+ *
+ * The handler creates exported document based on given options.
+ *
+ * @param {EventEmitter} emitter Instance of EventEmitter.
+ * @param {DatabaseManager} connectionInstance Instance of DatabaseManager.
+ *
+ * @returns {function} Event handler.
+ */
+module.exports = (emitter, connectionInstance) => {
```
Comment (tiko-star) [inline handlers/onImportConsumed.js:16]: ste normal function gri eli, es unnamednery menak runtimer hamar anenq

File: handlers/onImportConsumed.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+"use strict";
+
+const config = require("../config.js");
+const DocumentImporter = require("../service/DocumentImporter");
+
+/**
+ * Return event handler for emitted "importConsumed" events.
```
Comment (tiko-star) [inline handlers/onImportConsumed.js:7]: es anunnel chisht gri eventi

File: handlers/onImported.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Return event handler for emitted "imported" events.
+ *
+ * The handler produces notification about exported document.
+ *
+ * @param {number} compressionType Message compression type.
+ *
+ * @returns {function} Event handler.
+ */
+module.exports = (compressionType) => {
```
Comment (tiko-star) [inline handlers/onImported.js:14]: stexel normal function gri

File: errors/InvalidEntityTypeError.js [hunk 8]
```diff
@@ -1,6 +1,6 @@
 "use strict";
```
Comment (tiko-star) [inline errors/InvalidEntityTypeError.js:1]: es file service mej Error folder sarqi ytex pahi, ira levely service levela

File: service/DocumentCreator/ColumnResolver/index.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
 "use strict";
 
-const InvalidEntityTypeError = require("../Error/InvalidEntityTypeError.js");
+const InvalidEntityTypeError = require("../../../errors/InvalidEntityTypeError.js");
```
Comment (tiko-star) [inline service/DocumentCreator/ColumnResolver/index.js:3]: stexel Error - i mejic use kanes

File: service/DocumentCreator/Exporter.js [hunk 10]
```diff
@@ -17,6 +17,8 @@ class Exporter {
      * Export document based on given options.
      *
      * @param {object} type, format and criteria for exporting.
+     * @param format
```
Comment (tiko-star) [inline service/DocumentCreator/Exporter.js:20]: es sranc typernel nshi

File: service/DocumentImporter/DocumentImporter.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
+"use strict";
+
+const Factory = require("./Factory.js");
+const { join } = require("path");
```
Comment (tiko-star) [inline service/DocumentImporter/DocumentImporter.js:4]: ete usadir es exel es desturctov importnery nerqevum em grum sovoraknnery verevum

File: service/DocumentImporter/DocumentImporter.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+const { importsPath } = require("../../config");
+const createQueryResolver = require("./QueryResolver/QueryResolverFactory");
+const ResultReporter = require('./ResultReporter');
+
+class DocumentImporter {
+    constructor(connection) {
```
Comment (tiko-star) [inline service/DocumentImporter/DocumentImporter.js:10]: stex comment dir

File: service/DocumentImporter/DocumentImporter.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        this.connection = connection;
+    }
+
+    async importDocument(record) {
+        const { type, filename, projectId, companyId, mapping } = record;
+        const queryResolver = createQueryResolver(type, this.connection);
+        const resultReporter = new ResultReporter();
+        const importer = Factory.createImporter(type, this.connection, queryResolver, resultReporter);
```
Comment (tiko-star) [inline service/DocumentImporter/DocumentImporter.js:18]:
```js
const importer = Factory.createImporter(type, this.connection, queryResolver, new ResultReporter());
```
xi chenq senc anum?

File: service/DocumentImporter/DocumentImporter.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+const { join } = require("path");
+const { importsPath } = require("../../config");
+const createQueryResolver = require("./QueryResolver/QueryResolverFactory");
+const ResultReporter = require('./ResultReporter');
+
+class DocumentImporter {
+    constructor(connection) {
+        this.connection = connection;
+    }
+
```
Comment (tiko-star) [inline service/DocumentImporter/DocumentImporter.js:13]: comment

File: service/DocumentImporter/DocumentImporter.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        const pathname = this.generatePathname(projectId, filename);
+
+        return await importer.import({
+            type,
+            companyId,
+            pathname,
+            mapping,
+        });
+    }
+
```
Comment (tiko-star) [inline service/DocumentImporter/DocumentImporter.js:29]: comment

File: service/DocumentImporter/DocumentImporter.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
+"use strict";
+
+const Factory = require("./Factory.js");
+const { join } = require("path");
+const { importsPath } = require("../../config");
+const createQueryResolver = require("./QueryResolver/QueryResolverFactory");
+const ResultReporter = require('./ResultReporter');
+
```
Comment (tiko-star) [inline service/DocumentImporter/DocumentImporter.js:8]: comment

File: service/DocumentImporter/Driver/AbstractDriver.js [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        this.connection = connection;
+        this.queryResolver = queryResolver;
+    }
+
+    /**
+     * Creates a workbook from a file.
+     * After that, it creates worksheets depending on the type of the imported file.
+     *
+     * @param pathname
+     * @returns {Promise<Array>}
```
Comment (tiko-star) [inline service/DocumentImporter/Driver/AbstractDriver.js:37]: es type sxala

File: service/DocumentImporter/Driver/CategoryImporterDriver.js [hunk 18]
```diff
@@ -0,0 +1,356 @@
+"use strict";
+
+const AbstractDriver = require("./AbstractDriver");
+const Stack = require('../../../utils/Stack');
+const applyMapping = require('../MappingApplier');
+
```
Comment (tiko-star) [inline service/DocumentImporter/Driver/CategoryImporterDriver.js:6]: Poxos voch mi comment ches grel

File: service/DocumentImporter/Driver/CategoryImporterDriver.js [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+        for (const worksheet of worksheets) {
+            await this.resolveWorksheet({
+                worksheet,
+                companyId,
+                mapping,
+            });
+        }
+
+        return this.resultReporter;
```
Comment (tiko-star) [inline service/DocumentImporter/Driver/CategoryImporterDriver.js:25]: es side effecta?

File: service/DocumentImporter/Driver/CategoryImporterDriver.js [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+    }
+
+    async resolveWorksheet(
+        {
+            worksheet,
+            companyId,
+            mapping,
```
Comment (tiko-star) [inline service/DocumentImporter/Driver/CategoryImporterDriver.js:32]: stex itogum mapingy arandzin es hashvum?

File: service/DocumentImporter/QueryResolver/Resolvers/CategoryQueryResolver.js [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+
+        bindings.push(slug, slug);
+
+        const query = `INSERT INTO category_translations (${keys}) VALUES (${values})`;
+
+        return {query, bindings};
+    }
+
+    /**
+     * Updates given category parentId by category id.
```
Comment (tiko-star) [inline service/DocumentImporter/QueryResolver/Resolvers/CategoryQueryResolver.js:153]: Update

File: service/DocumentImporter/QueryResolver/Resolvers/CategoryQueryResolver.js [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+     *
+     * @param newParentId
+     * @param id
+     * @returns {{query: string, bindings: string[]}}
+     */
+    updateCategory({newParentId = null, id = null}) {
```
Comment (tiko-star) [inline service/DocumentImporter/QueryResolver/Resolvers/CategoryQueryResolver.js:159]: updateCategoryParentId, mekel xi en null by default? ete null ekan havaqi ban darnum

File: utils/Stack/index.js [hunk 23]
```diff
@@ -0,0 +1,43 @@
+"use strict";
+
+module.exports = class {
```
Comment (tiko-star) [inline utils/Stack/index.js:3]: normal class sarqi anunov

File: utils/Stack/index.js [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+	constructor(items = []) {
+		this.items = [
```
Comment (tiko-star) [inline utils/Stack/index.js:5]: qo mot itemsy arraya xies wrap anum arrayov?

File: utils/Stack/index.js [hunk 25]
```diff
@@ -0,0 +1,43 @@
+"use strict";
+
```
Comment (tiko-star) [inline utils/Stack/index.js:2]: es fileid formattingy sxala

**Review by tiko-star — CHANGES_REQUESTED** [review]

File: service/DocumentImporter/Driver/CategoryImporterDriver.js (hunk 20 above)
Comment (k3rnel) [reply service/DocumentImporter/Driver/CategoryImporterDriver.js:32]: ha, MappingApplier ka, iran talis es worksheetRow u mapping inqy xpuma irar u veradardznuma maqur row vorov karas ashxates

File: service/DocumentImporter/QueryResolver/Resolvers/CategoryQueryResolver.js (hunk 22 above)
Comment (k3rnel) [reply service/DocumentImporter/QueryResolver/Resolvers/CategoryQueryResolver.js:159]: esi toxel em updateCategory vortev es pahin parentina update anum, bayc heto status karoxa avelana, vor rename chanem methody. Commentum grel em es pahin incha anum es methody, heto kpoxem commenty erb vor avelacnenq status kam urish data

File: app.js [hunk 26]
```diff
@@ -31,12 +33,18 @@ const consumer = kafka.consumer(config.consumer);
 const emitter = new EventEmitter();
 const connectionInstance = DatabaseManager.getInstance("commerce");
 
-// After consuming start export of the document.
-emitter.on("consumed:export", onConsumed(emitter, connectionInstance));
+// After consuming start the export of the document.
+emitter.on("consumed:export", onExportConsumed(emitter, connectionInstance));
+
+// After consuming start the export of the document.
```
Comment (tiko-star) [inline app.js:39]: After consuming start the import of the document.

File: service/DocumentCreator/Driver/AbstractDriver.js [hunk 27]
```diff
@@ -42,7 +42,7 @@ class AbstractDriver {
     /**
      * Fetch required SQL query with bindings.
      *
-     * @param {string} type Exportable entity type.
+     * @param {Object} type Exportable entity type.
```
Comment (tiko-star) [inline service/DocumentCreator/Driver/AbstractDriver.js:45]: es String chi?

File: service/DocumentCreator/Exporter.js [hunk 28]
```diff
@@ -14,12 +14,13 @@ class Exporter {
     }
 
     /**
-     * Export document based on given options.
+     * Export document based on the given options.
      *
-     * @param {object} type, format and criteria for exporting.
+     * @param {object} type for exporting.
```
Comment (tiko-star) [inline service/DocumentCreator/Exporter.js:19]: es masery ber hin dzevin

**Review by tiko-star — APPROVED** [review]

**Pull request merged**